Airbyte based loaders

flash1293 opened the pull request:
This PR adds 8 new loaders:
* `AirbyteCDKLoader` This reader can wrap and run all python-based Airbyte source connectors.
* Separate loaders for the most commonly used APIs:
  * `AirbyteGongLoader`
  * `AirbyteHubspotLoader`
  * `AirbyteSalesforceLoader`
  * `AirbyteShopifyLoader`
  * `AirbyteStripeLoader`
  * `AirbyteTypeformLoader`
  * `AirbyteZendeskSupportLoader`

## Documentation and getting started
I added the basic shape of the config to the notebooks. This increases the maintenance effort a bit, but I think it's worth it to make sure people can get started quickly with these important connectors. This is also why I linked the spec and the documentation page in the readme as these two contain all the information to configure a source correctly (e.g. it won't suggest using oauth if that's avoidable even if the connector supports it).

## Document generation
The "documents" produced by these loaders won't have a text part (instead, all the record fields are put into the metadata). If a text is required by the use case, the caller needs to do custom transformation suitable for their use case.

## Incremental sync
All loaders support incremental syncs if the underlying streams support it. By storing the `last_state` from the reader instance away and passing it in when loading, it will only load updated records.

Commit:
wip

Commit:
add documentation and other connectors

baskaryan commented:
@flash1293 feel free to ping when this is ready for review

pedroslopez commented on docs/extras/integrations/document_loaders/airbyte_cdk.ipynb:
```diff
@@ -0,0 +1,222 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#!pip install \"source_github@git+https://github.com/airbytehq/airbyte.git@master#subdirectory=airbyte-integrations/connectors/source-github\""
```
Should these be updated to reference the PyPI connectors?

flash1293 replied:
No, for those they should use the other loaders, this is basically the "escape hatch"

pedroslopez commented on libs/langchain/langchain/document_loaders/airbyte_cdk.py:
```diff
@@ -0,0 +1,27 @@
+        return Document(page_content="", metadata=record.data)
+
+    def load(self) -> List[Document]:
+        return list(self._load_data(stream_name=self._stream_name, state=self._state))
```
nit: 

```suggestion
        return list(self.lazy_load())
```

pedroslopez commented on libs/langchain/langchain/document_loaders/airbyte_cdk.py:
```diff
@@ -0,0 +1,27 @@
+        return list(self._load_data(stream_name=self._stream_name, state=self._state))
+
+    def lazy_load(self) -> Iterator[Document]:
+        return self._load_data(stream_name=self._stream_name, state=self._state)
```
I notice we're passing state through... should the loader also internally update `_state` with the latest data, or do we want to make it explicit for users to pass through state themselves?

Just thinking about the ergonomics, someone that wants to implement incremental would need to instantiate a new instance of the loader if state changed. Not sure if instead we should pass state in the `load` method rather than on init.

Thinking about this from the Airbyte PoV, state changes on each run of the source, but the other info like configs/catalog stays the same throughout, which is why I'm leaning a bit towards moving state to the load.

pedroslopez replied:
Following the slack convo, if this is part of the langchain interface please ignore me 😛

pedroslopez commented on libs/langchain/langchain/document_loaders/airbyte_cdk.py:
```diff
@@ -0,0 +1,27 @@
+
+from langchain.docstore.document import Document
+from langchain.document_loaders.base import BaseLoader
+from airbyte_protocol.models.airbyte_protocol import AirbyteRecordMessage, AirbyteStateMessage
+from airbyte_cdk.sources.embedded.base_integration import BaseEmbeddedIntegration
+from airbyte_cdk.sources.embedded.runner import CDKRunner
```
Where are the `airbyte_protocol` and `airbyte_cdk` dependencies defined?

Do we expect users to install them separately?

flash1293 replied:
One is a transitive dependency of the other, but airbyte-cdk is actually re-exporting the types, so we can work with just airbyte_protocol here.

pedroslopez commented on libs/langchain/langchain/document_loaders/airbyte_cdk.py:
```diff
@@ -0,0 +1,27 @@
+    """Loads records using an Airbyte source connector implemented using the CDK."""
+
+    def __init__(self, config: Mapping[str, Any], source_class: Any, stream_name: str, state: Optional[AirbyteStateMessage] = None) -> None:
+        super().__init__(config=config, runner=CDKRunner(source=source_class(), name=source_class.__name__))
+        self._stream_name = stream_name
+        self._state = state
```
This might be something on the CDK runner side, but do we want to do any sort of validation on the config? I think if we're not doing so already, a quick win on UX could be to run `check` with the config on init

pedroslopez commented on docs/extras/integrations/document_loaders/airbyte_zendesk_support.ipynb:
```diff
@@ -0,0 +1,215 @@
+   "metadata": {},
+   "source": [
+    "Check out the [Airbyte documentation page](https://docs.airbyte.com/integrations/sources/zendesk-support/) for details about how to configure the reader.\n",
+    "The JSON schema the config object should adhere to can be found on Github: [https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-zendesk-support/source_zendesk-support/spec.json](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-zendesk-support/source_zendesk-support/spec.json).\n",
```
Looks like this link to the spec is a 404

pedroslopez commented on docs/extras/integrations/document_loaders/airbyte_zendesk_support.ipynb:
```diff
@@ -0,0 +1,215 @@
+    "    def __init__(self, config, state = None):\n",
+    "        super().__init__(config=config, stream_name=\"tickets\", state=state)\n",
+    "    \n",
+    "    def _handle_record(self, record, id):\n",
+    "        return Document(page_content=record.data[\"title\"], metadata=record.data)\n",
```
This might be an ok place to start, but I wonder if we can make this easier when creating the loader so don't have to extend the whole class.

I could see this as either a method passed on init, or exposing things like `document_text_field: "title"` and `document_metadata_fields: []` directly.

flash1293 replied:
Changed to a parameter passed into init

pedroslopez reviewed (commented):
Made some comments on some potential UX/ergonomic improvements, but generally looks good!

Thanks Joe 😄

Commit:
Merge remote-tracking branch 'upstream/master' into flash1293/airbyte-loader

Commit:
review comments

Commit:
review comments

Commit:
add gong and fix guarded import

flash1293 commented:
Thanks for the review @pedroslopez @eyurtsev  @hwchase17 , made some adjustments:
* Consolidated the loaders into a single file (`airbyte.cdk`). I played with having a single "PredefinedSourceLoader" with a string argument, but I think having separate classes is easier to use as your IDE can give you a list of them (and the boilerplate isn't crazy)
* Avoid importing anything on the top level
* Allow passing in a record_handler to build the document instead of inheritance (also adjusted the notebooks)
* Use the import guard to import the individual source to give a nice error message if they are not available
* Added gong as it was requested (happy to add others from https://docs.airbyte.com/integrations/ )

Things I'm not sure about:
* The record handler has an `AirbyteRecordMessage` argument, but how can I type that without importing? Not a blocker, but it would be nice to get rid of that `Any`

Commit:
adjust readme

baskaryan committed:
lint

baskaryan commented:
thanks @flash1293!

homanp commented:
@baskaryan

`from langchain.document_loaders.airbyte import AirbyteStripeLoader`

Throws: 
```
from libs.langchain.langchain.utils.utils import guard_import
ModuleNotFoundError: No module named 'libs'
```

aaronsteers commented:
> @baskaryan
> 
> `from langchain.document_loaders.airbyte import AirbyteStripeLoader`
> 
> Throws:
> 
> ```
> from libs.langchain.langchain.utils.utils import guard_import
> ModuleNotFoundError: No module named 'libs'
> ```

@homanp - Thanks for reporting. I am looking into this. 👀

eyurtsev commented:
A fix has been merged on master @aaronsteers

eyurtsev commented:
https://github.com/langchain-ai/langchain/pull/8998

aaronsteers commented:
Great - thanks, @eyurtsev ! 🎉

homanp commented:
@eyurtsev when will you release the next version?
